File: src/main/java/net/slipp/social/security/SlippDaoAuthenticationProvider.java
package net.slipp.social.security;

import net.slipp.domain.ProviderType;

import org.springframework.security.authentication.AuthenticationServiceException;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.authentication.dao.AbstractUserDetailsAuthenticationProvider;
import org.springframework.security.authentication.dao.SaltSource;
import org.springframework.security.authentication.encoding.PasswordEncoder;
import org.springframework.security.authentication.encoding.PlaintextPasswordEncoder;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.authority.mapping.GrantedAuthoritiesMapper;
import org.springframework.security.core.authority.mapping.NullAuthoritiesMapper;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.util.Assert;

public class SlippDaoAuthenticationProvider extends AbstractUserDetailsAuthenticationProvider {

    //~ Instance fields ================================================================================================

    private PasswordEncoder passwordEncoder = new PlaintextPasswordEncoder();

    private SaltSource saltSource;

    private UserDetailsService userDetailsService;
    
    private GrantedAuthoritiesMapper authoritiesMapper = new NullAuthoritiesMapper();

    //~ Methods ========================================================================================================

    @SuppressWarnings("deprecation")
    protected void additionalAuthenticationChecks(UserDetails userDetails,
            UsernamePasswordAuthenticationToken authentication) throws AuthenticationException {
        Object salt = null;

        if (this.saltSource != null) {
            salt = this.saltSource.getSalt(userDetails);
        }

        if (authentication.getCredentials() == null) {
            logger.debug("Authentication failed: no credentials provided");

            throw new BadCredentialsException(messages.getMessage("AbstractUserDetailsAuthenticationProvider.badCredentials", "Bad credentials"));
        }

        String presentedPassword = authentication.getCredentials().toString();

        if (!passwordEncoder.isPasswordValid(userDetails.getPassword(), presentedPassword, salt)) {
            logger.debug("Authentication failed: password does not match stored value");

            throw new BadCredentialsException(messages.getMessage("AbstractUserDetailsAuthenticationProvider.badCredentials", "Bad credentials"));
        }
    }
    
    protected Authentication createSuccessAuthentication(Object principal, Authentication authentication,
            UserDetails user) {
        UsernamePasswordAuthenticationToken result = new UsernamePasswordAuthenticationToken(principal,
                authentication.getCredentials(), authoritiesMapper.mapAuthorities(user.getAuthorities()));
        result.setDetails(ProviderType.slipp);

        return result;
    }
    
    protected void doAfterPropertiesSet() throws Exception {
        Assert.notNull(this.userDetailsService, "A UserDetailsService must be set");
    }

    protected final UserDetails retrieveUser(String email, UsernamePasswordAuthenticationToken authentication)
            throws AuthenticationException {
        UserDetails loadedUser;

        try {
            SlippUserDetailsService slippUserDetailsService = (SlippUserDetailsService)getUserDetailsService();
            loadedUser = slippUserDetailsService.loadUserByEmail(email);
        } catch (UsernameNotFoundException notFound) {
            throw notFound;
        } catch (Exception repositoryProblem) {
            throw new AuthenticationServiceException(repositoryProblem.getMessage(), repositoryProblem);
        }

        if (loadedUser == null) {
            throw new AuthenticationServiceException(
                    "UserDetailsService returned null, which is an interface contract violation");
        }
        return loadedUser;
    }

    /**
     * Sets the PasswordEncoder instance to be used to encode and validate passwords.
     * If not set, the password will be compared as plain text.
     * <p>
     * For systems which are already using salted password which are encoded with a previous release, the encoder
     * should be of type {@code org.springframework.security.authentication.encoding.PasswordEncoder}. Otherwise,
     * the recommended approach is to use {@code org.springframework.security.crypto.password.PasswordEncoder}.
     *
     * @param passwordEncoder must be an instance of one of the {@code PasswordEncoder} types.
     */
    public void setPasswordEncoder(Object passwordEncoder) {
        Assert.notNull(passwordEncoder, "passwordEncoder cannot be null");

        if (passwordEncoder instanceof PasswordEncoder) {
            this.passwordEncoder = (PasswordEncoder) passwordEncoder;
            return;
        }

        if (passwordEncoder instanceof org.springframework.security.crypto.password.PasswordEncoder) {
            final org.springframework.security.crypto.password.PasswordEncoder delegate =
                    (org.springframework.security.crypto.password.PasswordEncoder)passwordEncoder;
            this.passwordEncoder = new PasswordEncoder() {
                public String encodePassword(String rawPass, Object salt) {
                    checkSalt(salt);
                    return delegate.encode(rawPass);
                }

                public boolean isPasswordValid(String encPass, String rawPass, Object salt) {
                    checkSalt(salt);
                    return delegate.matches(rawPass, encPass);
                }

                private void checkSalt(Object salt) {
                    Assert.isNull(salt, "Salt value must be null when used with crypto module PasswordEncoder");
                }
            };

            return;
        }

        throw new IllegalArgumentException("passwordEncoder must be a PasswordEncoder instance");
    }

    protected PasswordEncoder getPasswordEncoder() {
        return passwordEncoder;
    }

    /**
     * The source of salts to use when decoding passwords. <code>null</code>
     * is a valid value, meaning the <code>DaoAuthenticationProvider</code>
     * will present <code>null</code> to the relevant <code>PasswordEncoder</code>.
     * <p>
     * Instead, it is recommended that you use an encoder which uses a random salt and combines it with
     * the password field. This is the default approach taken in the
     * {@code org.springframework.security.crypto.password} package.
     *
     * @param saltSource to use when attempting to decode passwords via the <code>PasswordEncoder</code>
     */
    public void setSaltSource(SaltSource saltSource) {
        this.saltSource = saltSource;
    }

    protected SaltSource getSaltSource() {
        return saltSource;
    }

    public void setUserDetailsService(UserDetailsService userDetailsService) {
        this.userDetailsService = userDetailsService;
    }

    protected UserDetailsService getUserDetailsService() {
        return userDetailsService;
    }
}

File: src/main/scala/net/slipp/social/security/SlippDaoAuthenticationProvider.scala
package net.slipp.social.security

import net.slipp.domain.ProviderType
import org.springframework.security.authentication.{AuthenticationServiceException, BadCredentialsException, UsernamePasswordAuthenticationToken}
import org.springframework.security.authentication.dao.AbstractUserDetailsAuthenticationProvider
import org.springframework.security.core.{Authentication, AuthenticationException}
import org.springframework.security.core.authority.mapping.{GrantedAuthoritiesMapper, NullAuthoritiesMapper}
import org.springframework.security.core.userdetails.{UserDetails, UserDetailsService, UsernameNotFoundException}
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.util.Assert

class SlippDaoAuthenticationProvider extends AbstractUserDetailsAuthenticationProvider {
  private var passwordEncoder: PasswordEncoder = null
  private var userDetailsService: UserDetailsService = null
  private var authoritiesMapper: GrantedAuthoritiesMapper = new NullAuthoritiesMapper

  protected def additionalAuthenticationChecks(userDetails: UserDetails, authentication: UsernamePasswordAuthenticationToken) {
    if (authentication.getCredentials == null) {
      logger.debug("Authentication failed: no credentials provided")
      throw new BadCredentialsException(messages.getMessage("AbstractUserDetailsAuthenticationProvider.badCredentials", "Bad credentials"))
    }
    val presentedPassword: String = authentication.getCredentials.toString
    if (!passwordEncoder.matches(presentedPassword, userDetails.getPassword)) {
      logger.debug("Authentication failed: password does not match stored value")
      throw new BadCredentialsException(messages.getMessage("AbstractUserDetailsAuthenticationProvider.badCredentials", "Bad credentials"))
    }
  }

  protected override def createSuccessAuthentication(principal: AnyRef, authentication: Authentication, user: UserDetails): Authentication = {
    val result: UsernamePasswordAuthenticationToken = new UsernamePasswordAuthenticationToken(principal, authentication.getCredentials, authoritiesMapper.mapAuthorities(user.getAuthorities))
    result.setDetails(ProviderType.slipp)
    result
  }

  protected override def doAfterPropertiesSet {
    Assert.notNull(this.userDetailsService, "A UserDetailsService must be set")
  }

  protected final def retrieveUser(email: String, authentication: UsernamePasswordAuthenticationToken): UserDetails = {
    var loadedUser: UserDetails = null
    try {
      val slippUserDetailsService: SlippUserDetailsService = getUserDetailsService.asInstanceOf[SlippUserDetailsService]
      loadedUser = slippUserDetailsService.loadUserByEmail(email)
    }
    catch {
      case notFound: UsernameNotFoundException => {
        throw notFound
      }
      case repositoryProblem: Exception => {
        throw new AuthenticationServiceException(repositoryProblem.getMessage, repositoryProblem)
      }
    }
    if (loadedUser == null) {
      throw new AuthenticationServiceException("UserDetailsService returned null, which is an interface contract violation")
    }
    return loadedUser
  }

  def setPasswordEncoder(passwordEncoder: PasswordEncoder) {
    Assert.notNull(passwordEncoder, "passwordEncoder cannot be null")
    this.passwordEncoder = passwordEncoder
  }

  protected def getPasswordEncoder: PasswordEncoder = {
    return passwordEncoder
  }

  def setUserDetailsService(userDetailsService: UserDetailsService) {
    this.userDetailsService = userDetailsService
  }

  protected def getUserDetailsService: UserDetailsService = {
    return userDetailsService
  }
}